### PR TITLE
Added a Bull client, and a route /feeds/info to query queue's info

### LINKED ISSUE #2414: Expose basic feed queue stats from Status service
In #2395 we are adding an expanded status service dashboard. One thing I want to include in it is the ability to see the current status of the feed queue.  That includes things like:

- how many jobs are in the queue?
- how many are active?
- how many have failed?
- how many are delayed?
- how many are waiting?

We use [Bull + Redis](https://github.com/OptimalBits/bull) to manage our [feed queue](https://github.com/Seneca-CDOT/telescope/blob/master/src/backend/feed/queue.js).  Bull has a nice [API for querying the status for a queue](https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md).

Let's add a new route to the Status service that returns details about the current state of the feed queue.

```
GET /feeds
```

It can return a complex JS object with various details queried from the queue.

### PR DESCRIPTION
**Pull request by TueNguyen2911**
## Issue This PR Addresses
This fixes #2414 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This is draft PR for issue #2414. 

I added a new file `queue.js` that creates and exports a new Bull client with `redis` client connection required from [posts/src/redis.js](https://github.com/Seneca-CDOT/telescope/blob/master/src/api/posts/src/redis.js). I also added a new GET route `feeds/info` to return the queue's information, for now I'm trying to return the number of jobs in the queue. 

I tested the route by running `npm run services:start`, `npm start` first and running `http://localhost/v1/posts/feeds/info` in Postman. However, I haven't got the expected response, please let me know if I'm on the right track!

The response I got was a long loading screen until Bad Gateway

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/api/posts/src/routes/feeds.js [hunk 1]
```diff
@@ -34,6 +35,18 @@ feeds.get('/', async (req, res, next) => {
...
+//This should return number of jobs in the queue, active, failed, delayed, waiting jobs
+feeds.get('/info', async (req, res, next) => {
+  let jobNum;
+  try {
+    jobNum = await queue.count();
+  } catch (err) {
+    logger.error({ error }, "Unable to get queue's info");
+    next(error);
+  }
+  res.json({ jobNum: jobNum });
```
Comment (humphd) [inline src/api/posts/src/routes/feeds.js:47]: Move this to after line 42, inside the try/catch.  Also, you don't need to repeat `jobNum`:

```js
res.json({ jobNum });
```

File: src/api/posts/src/routes/feeds.js [hunk 2]
```diff
@@ -42,7 +55,7 @@ feeds.get('/:id', validateFeedsIdParam(), async (req, res, next) => {
     const feed = await Feed.byId(id);
     if (!feed) {
       res.status(404).json({
-        message: `Feed not found for id ${id}`,
+        message: `Feed not found fore id ${id}`,
```
Comment (humphd) [inline src/api/posts/src/routes/feeds.js:58]: Spelling mistake.  Should be `for` not `fore`.

**Review by humphd — CHANGES_REQUESTED** [review]
You should try running this in GitPod so you don't have to set everything up locally.

Also, you have some lint errors to fix.

File: src/api/posts/src/routes/feeds.js [hunk 3]
```diff
@@ -34,6 +35,18 @@ feeds.get('/', async (req, res, next) => {
   );
 });
 
+// This should return number of jobs in the queue, active, failed, delayed, waiting jobs
+feeds.get('/info', async (req, res, next) => {
+  let jobNum;
+  try {
+    jobNum = await queue.count();
```
Comment (humphd) [inline src/api/posts/src/routes/feeds.js:42]: Tested on GitPod and this route never seems to return (e.g., I get no response from the server, it just loads forever).  It looks like you're doing it correctly, but there's nothing coming back.

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by TueNguyen2911** [comment]
That's what I got for my response as well, the idea seems correct though. I'll look more into it

**Comment by humphd** [comment]
I suspect that the connection to Redis isn't right somehow (e.g., it's never returning a response).

**Comment by TueNguyen2911** [comment]
I'll compare it to the connection in `backend`. Is it possible to log the connection using `logger`?

**Comment by TueNguyen2911** [comment]
Hi @humphd, it's not returning any data because the redis clients from `backend` and `posts` have different host. 

The `Redis_URL` in `backend` is `redis:://127.0.0.1` while the one from `satellite`  is `redis://redis:6379`
```js
const redisUrl = process.env.REDIS_URL || 'redis://redis:6379'; // redis.js from satellite
```
https://github.com/Seneca-CDOT/telescope/blob/2359565d11aa045eedadfee2ef72d29f62d87bee/src/backend/lib/redis.js#L9

I tried uncommenting this line ` #REDIS_URL=redis://127.0.0.1 ` from [env.development](https://github.com/Seneca-CDOT/telescope/blob/master/config/env.development) and building the containers again using `docker-compose --env-file ./config/env.development up --build`  but the `host` remain the same.

How do I change the variable `REDIS_URL` to `redis:://127.0.0.1` for satellite's redis.js file

**Comment by TueNguyen2911** [comment]
Thank you! I tested it out and the queue from `api` successfully returned stats data. 

If cleaning up the backend is on the list for the next release, it'll be probably late for my release 0.4. Also, I don't think I should touch the configuration of `backend` in this PR (maybe in another PR). 

So how should I move forward from here, should I wait for the other devs' ideas?

**Comment by manekenpix** [comment]
I'm not sure if this is a `redis` issue since all the other endpoints in `api/posts` rely on the same `redis` client used in the `feeds/info` endpoint, and they all can connect and retrieve data from `redis`.

**Comment by humphd** [comment]
I agree it's odd that this isn't working, since everything else is running against that same Redis instance.  Is it just your dev setup?

**Comment by TueNguyen2911** [comment]
I agree, if it works for other endpoints why not `info`?

It's just when I followed Josue's instructions, the promise from `queue.count()` got resolved once, when `/info` was hit, it still didn't return nothing. I thought changing the `.env` would solve the problem. 

I'll keep trying.

I'll look more into it.

**Comment by TueNguyen2911** [comment]
Turns out, I need to provide the `createClient` with type `client` or `subscriber` for queues to share the same connection according to [the documentation](https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md#queue). 

This is the response I've got for now, `completed` is also 0
```json
{"queueInfo":{"waiting":785,"active":0,"completed":0,"failed":2,"delayed":0,"paused":0,"jobCnt":785}}
```

The reason `completed` is 0 is because of the 2 properties below are true.
https://github.com/Seneca-CDOT/telescope/blob/4b1c07464e79507b29fcabf7a91b3ffafa0068c2/src/backend/feed/queue.js#L29-L30

I also added the 2 options `const redis = Redis({ maxRetriesPerRequest: null, enableReadyCheck: false });` in `redis.js` because I ran into this problem when building the containers. The [CHANGELOG](https://github.com/OptimalBits/bull/blob/develop/CHANGELOG.md#400-2021-10-27) from Bull also mentions this 

docker-error

File: src/api/posts/src/redis.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 const { Redis } = require('@senecacdot/satellite');
 
-const redis = Redis();
+const redis = Redis({ maxRetriesPerRequest: null, enableReadyCheck: false });
```
Comment (humphd) [inline src/api/posts/src/redis.js:3]: Why these options?

File: src/api/posts/src/routes/feeds.js [hunk 5]
```diff
@@ -28,6 +29,18 @@ feeds.get('/', async (req, res, next) => {
   );
 });
 
+feeds.get('/info', async (req, res, next) => {
+  try {
+    const jobCnt = await queue.count();
```
Comment (humphd) [inline src/api/posts/src/routes/feeds.js:34]: You could run these at the same time:

```js
const [jobCnt, queueInfo] = await Promise.all([queue.count(), queue.getJobCounts()]);
```

**Review by humphd — CHANGES_REQUESTED** [review]
We need to add a test for this as well in https://github.com/Seneca-CDOT/telescope/blob/master/src/api/posts/test/posts.test.js.  We won't have real data, but hopefully it should return something.

I'm fine if you do that test here or in a follow-up PR.  If you do the latter, please file the issue now.

**Comment by TueNguyen2911** [comment]
> We need to add a test for this as well in https://github.com/Seneca-CDOT/telescope/blob/master/src/api/posts/test/posts.test.js. We won't have real data, but hopefully it should return something.
> 
> I'm fine if you do that test here or in a follow-up PR. If you do the latter, please file the issue now.

I'll add tests in this PR

**Comment by TueNguyen2911** [comment]
Hi, for now, `waiting`, `active`, `delayed`, `jobCnt` of the response updates on changes of the queue. `completed`, `failed` are always 0 as completed and failed jobs are removed from the queue. 
```json
{"queueInfo":{"waiting":785,"active":0,"completed":0,"failed":2,"delayed":0,"paused":0,"jobCnt":785}}
```

I try listening for completed queue events but the event is never fired as the `queue` in post is not consumer or producer. I also try adding `'global:completed'`, the event is fired and I've got expected response but it returns error when I try to use `getJobcounts()` on the queue. The error is "Connection in subscriber mode, only subscriber commands may be used"

```javascript
//this event is never fired
queue.on('completed', (job) => {
  console.log(`${job.id} completed`);
});
//this event is fired but get error when trying to use Bull Api on queue
queue.on('global:completed', (jobid) => {
  console.log(`${jobid} completed`);
});
```
I'm not sure how to get `completed` and `active` to update. I need some help on this
Helpful doc links: [bull](https://optimalbits.github.io/bull/), [Global Events](https://github.com/OptimalBits/bull/blob/HEAD/REFERENCE.md#global-events)

**Comment by humphd** [comment]
This is one reason why doing this in the parser would make the most sense, since that's where we actually manage the queue to begin with.

That said, why don't we start with the items you can get, and do other things in follow ups?

**Comment by TueNguyen2911** [comment]
If that's the case, I'd love to work on the parser service later on. 

I'll follow up with what I can get right now. Thanks.

**Review by humphd — APPROVED** [review]

**Review by humphd — DISMISSED** [review]

**Comment by TueNguyen2911** [comment]
I clicked on Update Branch accidentally.

**Comment by humphd** [comment]
```
git checkout -B issue-2414 2fd0b98b1090925ca2869cf7a3dbabb97cf770fc
git checkout master
git pull upstream master
git checkout issue-2414
git rebase master
git push origin issue-2414 -f
```

**Comment by TueNguyen2911** [comment]
After this got merged, I need to work on the front-end right?

**Comment by humphd** [comment]
Yup, you can start now if you want, just base your new branch on this branch, and rebase onto master when it lands.

**Review by humphd — DISMISSED** [review]

**Comment by TueNguyen2911** [comment]
Got it!

**Review by HyperTHD — DISMISSED** [review]
Just for reference, Redis was added to Satellite so we should extract it from there. That way, we can also remove `ioredis` and `ioredis-mock` from the posts service. We can do that in another issue though. Nice job

**Comment by humphd** [comment]
@TueNguyen2911 let's get this in soon, can you rebase and fix the conflict in your package.json?  Ping me when it's ready so I can merge.

File: src/api/posts/package.json [hunk 6]
```diff
@@ -17,6 +17,7 @@
   },
   "dependencies": {
     "@senecacdot/satellite": "^1.x",
+    "bull": "^3.20.1",
```
Comment (humphd) [inline src/api/posts/package.json:20]: Because you a dependency, you need to also update the `pnpm` lock file and commit that as well.

**Review by humphd — CHANGES_REQUESTED** [review]

**Pull request merged**